lib/modules/manager/custom/jsonata/utils.ts

```diff
@@ -0,0 +1,119 @@
+import is from '@sindresorhus/is';
+import jsonata from 'jsonata';
+import { logger } from '../../../../logger';
+import * as template from '../../../../util/template';
+import { parseUrl } from '../../../../util/url';
+import type { PackageDependency } from '../../types';
+import { QueryResultZodSchema } from './schema';
+import type { JSONataManagerTemplates, JsonataExtractConfig } from './types';
+
+export const validMatchFields = [
+  'depName',
+  'packageName',
+  'currentValue',
+  'currentDigest',
+  'datasource',
+  'versioning',
+  'extractVersion',
+  'registryUrl',
+  'depType',
+] as const;
+
+type ValidMatchFields = (typeof validMatchFields)[number];
+
+export async function handleMatching(
+  json: unknown,
+  packageFile: string,
+  config: JsonataExtractConfig,
+): Promise<PackageDependency[]> {
+  // Pre-compile all JSONata expressions once
+  const compiledExpressions = config.matchStrings
+    .map((query) => {
+      try {
+        return jsonata(query);
+      } catch (err) {
+        logger.warn({ query, err }, 'Failed to compile JSONata query');
+        return null;
+      }
+    })
+    .filter((expr) => expr !== null);
+
+  // Execute all expressions in parallel
+  const results = await Promise.all(
+    compiledExpressions.map(async (expr) => {
+      try {
+        // can either be a single object, an array of objects or undefined (no match)
+        let result = (await expr.evaluate(json)) ?? [];
+        if (is.emptyObject(result) || is.emptyArray(result)) {
+          return [];
+        }
+
+        result = is.array(result) ? result : [result];
+
+        QueryResultZodSchema.parse(result);
+        return structuredClone(result);
```

Comment:
```suggestion
        const parsed = QueryResultZodSchema.safeParse(result);
        if ( parsed.success ) {
          return structuredClone(parsed.data);
        }
        logger.warn({ err: parsed.error, result }, 'Returned value does not meet minimum required fields');
        return []
```

Reply:
We cannot perform the check for minimum required fields using the Zod Schema here, because we still need to process the templates. I think having a zod schema here does not makes much sense.

Reply:
As discussed in the 1:1 chat, this should be still there with a different log message to safeguard against false types.